### PR TITLE
Refactor impl format

### PR DESCRIPTION
**Pull request by topecongiro**
This PR updates formatting against long impls.
Passes all stress tests from #1357 and #1369. ~~when using `rfc-rustfmt.toml`, but fails on some tests when using the current default configutation.~~
~~I am creating this PR anyway since we will be switching the default to Rfc style and this PR does cause no regression.~~

### CODE REVIEW

**Comment by topecongiro** [comment]
Added commits to support the current default setting.

**Comment by topecongiro** [comment]
Added commits to close #1168 and close #1642 and rebased.

File: src/items.rs [hunk 1]
```diff
@@ -790,6 +737,55 @@ fn format_impl_ref_and_type(context: &RewriteContext,
     }
 }
 
+// Returns false if failed to update result: then, try using multiline.
+fn format_trait_ref_then_update_result(context: &RewriteContext,
```
Comment (nrc) [inline src/items.rs]: Why update result rather than taking an offset and returning a substring (well, an Option)?

Comment (topecongiro) [inline src/items.rs]: Thanks for your feedback!
No real reason, I was being lazy here.
I would like to change this to return `Option<String>`.

**Comment by topecongiro** [comment]
Added a commit to rename `format_trait_ref_then_update_result` to `rewrite_trait_ref` and to change its return type to `Option<String>`.

**Comment by nrc** [comment]
Thanks for the change!

**Pull request merged**